File: carp.dsp.core/src/jvmTest/resources/scripts/load_data.py
#!/usr/bin/env python
"""
Simple deterministic data loading script.
Reads input CSV, validates it, writes output CSV.
"""

import sys
import csv
from pathlib import Path

def load_data(input_file: str, output_file: str) -> None:
    """
    Load and validate CSV file.
    
    Args:
        input_file: Path to input CSV
        output_file: Path to output CSV
    """
    input_path = Path(input_file)
    output_path = Path(output_file)
    
    # Create parent directories if needed
    output_path.parent.mkdir(parents=True, exist_ok=True)
    
    # If input doesn't exist, create minimal output
    if not input_path.exists():
        print(f"Warning: Input file {input_file} not found. Creating minimal output.", file=sys.stderr)
        with open(output_path, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'value'])
            writer.writerow([1, 10])
        return
    
    # Read input and load
    try:
        with open(input_path, 'r', newline='') as infile:
            reader = csv.DictReader(infile)
            rows = list(reader)
        
        # Validate: ensure no empty rows
        valid_rows = [row for row in rows if any(row.values())]
        
        # Get fieldnames
        if valid_rows:
            fieldnames = list(valid_rows[0].keys())
        else:
            fieldnames = ['id', 'value']
            valid_rows = [{'id': '1', 'value': '10'}]
        
        # Write output (copy of input with validation)
        with open(output_path, 'w', newline='') as outfile:
            writer = csv.DictWriter(outfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(valid_rows)
        
        print(f"Loaded and validated {len(valid_rows)} rows from {input_file} to {output_file}")
        
    except Exception as e:
        print(f"Error loading data: {e}", file=sys.stderr)
        # Create minimal output on error
        with open(output_path, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'value'])
            writer.writerow([1, 10])

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <input_file> <output_file>")
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2]
    
    load_data(input_file, output_file)
    print(f"Successfully loaded {input_file} → {output_file}")

File: carp.dsp.core/src/jvmTest/resources/scripts/load_data.R
#!/usr/bin/env Rscript
# Simple deterministic data loading script
# Reads input CSV, validates it, writes output CSV

# Get command-line arguments
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("Usage: load_data.R <input_file> <output_file>")
}

input_file <- args[1]
output_file <- args[2]

# Create output directory if needed
dir.create(dirname(output_file), showWarnings = FALSE, recursive = TRUE)

tryCatch({
  # Read input CSV
  if (file.exists(input_file)) {
    data <- read.csv(input_file, stringsAsFactors = FALSE)
    message(sprintf("Loaded %d rows from %s", nrow(data), input_file))
  } else {
    message(sprintf("Warning: Input file %s not found. Creating minimal output.", input_file))
    data <- data.frame(id = 1, value = 10)
  }
  
  # Validate: remove rows with all NA/empty values
  data <- data[rowSums(is.na(data) | data == "") < ncol(data), ]
  
  # Ensure we have at least one row
  if (nrow(data) == 0) {
    data <- data.frame(id = 1, value = 10)
  }
  
  # Write output CSV
  write.csv(data, output_file, row.names = FALSE)
  message(sprintf("Validated and loaded %d rows to %s", nrow(data), output_file))
  message(sprintf("Successfully loaded %s → %s", input_file, output_file))
  
}, error = function(e) {
  message(sprintf("Error loading data: %s", conditionMessage(e)))
  # Create minimal output on error
  data <- data.frame(id = 1, value = 10)
  write.csv(data, output_file, row.names = FALSE)
  message(sprintf("Created minimal output: %s", output_file))
})
